### PR TITLE
Name table stores with unique identifier to prevent having to rename them

### PR DESCRIPTION
**Pull request by yaziine**
This is a first iteration for solving #95. The schema of the DB has been changed, a new way of referencing tables has been introduced. 

Before, the store `_genji.tables` contained the configuration of each table and each store were keyed by the name of the table. But now, `_genji.tables` still holds the configuration of the table but also a `storeID` that contains a generated random value that will be the key of the store which contains the content of the table. This change will allow us to easily rename a table without duplicating its content (see https://github.com/genjidb/genji/issues/95#issuecomment-652399206 for more details).

Major changes:
- `storeID` introduced
- functions of the `Transaction` interface receive a `[]byte` now
- `ListStores` has been removed from the `Transaction` interface
- package `engineutil` has been removed, I removed because the implementation wasn't up to date anymore, we can add it later if needed

### CODE REVIEW

File: database/config.go [hunk 1]
```diff
@@ -134,46 +139,112 @@ func (f *FieldConstraint) ScanDocument(d document.Document) error {
...
+	ti.cfg = &TableConfig{}
+	return ti.cfg.ScanDocument(doc)
 }
 
-func (t *tableConfigStore) Replace(tableName string, cfg *TableConfig) error {
+type tableInfoStore struct {
+	st engine.Store
+}
+
+// Insert inserts a new tableInfo for the table tableName.
```
Comment (asdine) [inline database/config.go:188]:
```suggestion
// Insert a new tableInfo for the given table name.
```

File: database/config.go [hunk 2]
```diff
@@ -134,46 +139,112 @@ func (f *FieldConstraint) ScanDocument(d document.Document) error {
...
+// If the generated storeID already exists, a new one will be generated
```
Comment (asdine) [inline database/config.go:189]:
```suggestion
// It automatically generates a unique storeID for that table.
```

File: database/config.go [hunk 3]
```diff
@@ -134,46 +139,112 @@ func (f *FieldConstraint) ScanDocument(d document.Document) error {
...
+// until it doesn't already exist.
```
Comment (asdine) [inline database/config.go:190]:
```suggestion
```

File: database/config.go [hunk 4]
```diff
@@ -134,46 +139,112 @@ func (f *FieldConstraint) ScanDocument(d document.Document) error {
...
 
-type tableConfigStore struct {
-	st engine.Store
+type tableInfo struct {
+	// storeID is a generated ID that acts as a key to reference this table.
+	// The first-4 bytes represents the timestamp in second and the last-2 bytes are
+	// randomly generated.
+	storeID [6]byte
+
+	cfg *TableConfig
```
Comment (asdine) [inline database/config.go:148]: Why do we need a separate `TableConfig` type? What about just keeping everything in the same structure?

```go
type TableInfo struct {
  StoreID [6]byte
  FieldConstraints []FieldConstraint
  LastKey int64
}
```

File: database/config.go [hunk 5]
```diff
@@ -202,6 +272,28 @@ func (t *tableConfigStore) Delete(tableName string) error {
 	return err
 }
 
+func (t *tableInfoStore) ListTables() ([]string, error) {
+	it := t.st.NewIterator(engine.IteratorConfig{Reverse: false})
```
Comment (asdine) [inline database/config.go:276]: 🌷 It's reverse by default
```suggestion
	it := t.st.NewIterator(engine.IteratorConfig{})
```

File: database/table.go [hunk 6]
```diff
@@ -421,14 +422,14 @@ func (t *Table) Truncate() error {
 	return t.Store.Truncate()
 }
 
-// TableName returns the name of the table.
-func (t *Table) TableName() string {
+// Name returns the name of the table.
```
Comment (asdine) [inline database/table.go:425]: 👍

File: database/transaction.go [hunk 7]
```diff
@@ -12,19 +11,20 @@ import (
...
 type Transaction struct {
-	db         *Database
-	Tx         engine.Transaction
-	writable   bool
-	tcfgStore  *tableConfigStore
+	db       *Database
+	Tx       engine.Transaction
+	writable bool
+
+	infoStore  *tableInfoStore
```
Comment (asdine) [inline database/transaction.go:27]: We should be explicit here

```suggestion
	tableInfoStore  *tableInfoStore
```

File: index/index.go [hunk 8]
```diff
@@ -97,14 +97,19 @@ func NewUniqueIndex(tx engine.Transaction, idxName string) *UniqueIndex {
...
-	var b strings.Builder
-	b.WriteString(StorePrefix)
-	b.WriteString(name)
-	b.WriteByte(separator)
-	b.WriteByte(byte(t))
+func buildIndexName(name []byte, t Type) []byte {
+	var buf bytes.Buffer
 
-	return b.String()
+	// We can deterministically set the size of the buffer.
```
Comment (asdine) [inline index/index.go:103]: 👏

File: database/config.go (hunk 4 above)
Comment (yaziine) [reply database/config.go:148]: This is a good point. There is no strict reason behind this. I'll do that 👍

Comment (yaziine) [reply database/config.go:148]: After reflexion, this change will imply a loooot of diff. I suggest to merge this one, and I'll do that later on in a dedicated PR.

**Pull request merged**